### ISSUE TITLE
parallelism not honored

### ISSUE THREAD

**Issue by binnsr** [opentofu/opentofu#1678]
### OpenTofu Version

```shell
$ tofu version
OpenTofu v1.7.1
on linux_amd64
+ provider registry.opentofu.org/betterstackhq/logtail v0.1.14
+ provider registry.opentofu.org/cloudflare/cloudflare v4.13.0
+ provider registry.opentofu.org/digitalocean/digitalocean v2.36.0
+ provider registry.opentofu.org/dopplerhq/doppler v1.2.4
+ provider registry.opentofu.org/gitlabhq/gitlab v16.3.0
+ provider registry.opentofu.org/hashicorp/external v2.3.3
+ provider registry.opentofu.org/hashicorp/helm v2.13.2
+ provider registry.opentofu.org/hashicorp/kubernetes v2.30.0
+ provider registry.opentofu.org/hashicorp/random v3.6.2
+ provider registry.opentofu.org/upstash/upstash v1.5.3
```

### OpenTofu Configuration Files

```hcl
resource "upstash_kafka_topic" "topic" {
    for_each = { for topic in var.cluster_topics : topic.topic_name => topic }

    topic_name       = "${ each.key }"
    partitions       = "${ each.value.partitions }"
    retention_time   = "${ each.value.retention_time }"
    retention_size   = "${ each.value.retention_size }"
    max_message_size = "${ each.value.max_message_size }"
    cleanup_policy   = "${ each.value.cleanup_policy }"
    cluster_id       = upstash_kafka_cluster.cluster.cluster_id
}

    cluster_topics = [
        {
            topic_name             = "example"
            partitions             = 1
            retention_time         = 1800000
            retention_size         = 725124
            max_message_size       = 829213
            cleanup_policy         = "delete"
        },
.....
        {
            topic_name             = "example_200"
            partitions             = 1
            retention_time         = 1800000
            retention_size         = 725124
            max_message_size       = 829213
            cleanup_policy         = "delete"
        }
}
```

### Debug Output

https://gist.github.com/binnsr/f05ba87e7ea6683f33ba67e503b85e93

### Expected Behavior

Successful plan and apply of the kafka topics to upstash

### Actual Behavior

752 instances of '429 Too Many Requests' returned and a failed plan.
This works on opentofu 1.6.0-alpha5, which we've been using to work around the issue.
1.6.0-alpha5 fails if `-parallelism=1` is not set.
anything newer fails even if `-parallelism=1` is set.

### Steps to Reproduce

tofu init
tofu plan
fail

**Comment by ollevche** [comment]
Hey @binnsr, thank you for your contribution!

Parallelism flag is implemented with no ties to actual resources. So I checked it with a simpler [HTTP data resource](https://library.tf/providers/hashicorp/http/latest/docs/data-sources/http) and it produced the expected results (RPS to local resource was matching parallel setting).

In your case, many moving parts are touched, so it's hard to track why you get "Too Many Requests" from kafka. Can you provide a more complete (and isolated) example of the **test** setup (full .tf configuration, kafka settings, etc)?

**Comment by binnsr** [comment]
@ollevche Thanks for the update -- This is the exact error that tofu returns while refreshing state during a plan:
```
│ Error: Get Kafka Topic failed, status code: 429 response: <html>
│ <head><title>429 Too Many Requests</title></head>
│ <body>
│ <center><h1>429 Too Many Requests</h1></center>
│ <hr><center>nginx</center>
│ </body>
│ </html>
│ 
│ 
│   with module.v1.module.dev_cluster.upstash_kafka_topic.topic["user.setup"],
│   on ../../../modules/common/kafka/1.0.0/main.tf line 17, in resource "upstash_kafka_topic" "topic":
│   17: resource "upstash_kafka_topic" "topic" {
```
The Upstash API has rate-limits that we reach with both our prod and preprod environments.  On tofu 1.6.0-alpha5, adding `-parallelism=1` gets us past the limiting issue, but on subsequent versions, it still hits the rate limit.  Per upstash support, they cannot (or will not) change the rate-limit.
here's a [gist](https://gist.github.com/binnsr/542d235db4a16be88a77462bb0da09cb) with the module and a workspace that replicates the behavior. use only 'main.tf' initially, then add 'main-2.tf' .. with only main.tf, it's hit or miss whether you hit the rate limit.

**Comment by binnsr** [comment]
They've not said how they are rate limiting - a coworker has been working with them on this issue ( https://github.com/upstash/terraform-provider-upstash/issues/40 )
They are looking into ways to update their provider to resolve the issue, but there's been no traffic on the Issue since February 2024.

**Comment by ollevche** [comment]
Hey @binnsr, thank you for providing the gist. I will check it out and share my thoughts.

Also, to extend @janosdebugs answer, it makes sense to ensure that tofu is the only one making requests to the API since rate limiting could be calculated based on a lot of different resources.

**Comment by binnsr** [comment]
We're in the process of updating our codebase to side-step the issue by using different API keys per environment, but it still doesn't resolve why we can use `-parallelism=1` successfully in 1.6.0-alpha5 and before, but not in subsequent versions.

**Comment by ollevche** [comment]
Hey @binnsr, I've tested configuration provided in [the gist](https://gist.github.com/binnsr/542d235db4a16be88a77462bb0da09cb). Thanks for preparing this!

I did run different plan/apply scenarios and I can confirm tofu calls provider sequentially with `parallelism=1`.

The same behaviour is also seen in trace logs you provided. Let's compare timestamps of multiple `ReadResource` invocations: the delay is ~200 ms for `v1.7.2` and ~300 ms for `v1.6.0-alpha5`.

Running tofu with the testcase on my local machine doesn't help with reproducing the `Too Many Requests` since the delay between calls is closer to 500 ms. Upstash's API is located in US and I'm in Europe so there is additional latency. To validate the assumption, I also run the same testcase from US server and the apply failed even with the first main (103 resources, delay between `ApplyResourceChange` calls is under 200 ms).

The only question is why it works with the `v1.6.0-alpha5`.

As @janosdebugs pointed out earlier, it's possible to hit the rate limiting even without multithreading if you do sequential requests fast enough.

I believe tofu `v1.7.2` is more efficient in terms of graph traversal and/or making external requests then the older version (`v1.6.0-alpha5`). Combined with being on the edge of Upstash's rate limits, it results in occassional failures.

I think the proper fix should be implemented in the upstash provider to follow the rules of their interal rate limiting logic. Configurable max RPS setting could also help.
